### PR TITLE
Connection reuse support

### LINKED ISSUE #265: Connection Reuse Support
As an Agent Framework Interop Tester I want the AATH tests to reuse connections that have been established between the two instances of the roles in the protocol so that agents that support reuse can be tested and so that agents that don't support multiple connections can successfully run tests.

### PR DESCRIPTION
**Pull request by nodlesh**
Details on how to use connection reuse in AATH is located in CONNECTION-REUSE.md.

Closes #265

### CODE REVIEW

File: aries-backchannels/backchannel_operations.csv [hunk 1]
```diff
@@ -9,6 +9,7 @@ rfc,command,response,topic,method,operation,id,data,description,id_details,data_
 0160 Connection Protocol,X,,connection,POST,send-ping,Y,Y,Acknowledge acceptance of connectionr request,connection_id,ack,"connection_id, state"
 0160 Connection Protocol,,X,connection,GET,,Y,Y,Receive problem report,connection_id,ack,"connection_id, state"
 0160 Connection Protocol,X,,connection,POST,,Y,Y,Reject connection request and send message,connection_id,rejection,"connection_id, state"
+0160 Connection Protocol,X,,active-connection,GET,,Y,,Retrieve an active connection based on the did of the other role,thier_did,,"connection_id, my_did"
```
Comment (swcurran) [inline aries-backchannels/backchannel_operations.csv:12]: Typo "their_did"?

File: aries-test-harness/features/0023-did-exchange.feature [hunk 2]
```diff
@@ -151,4 +151,20 @@ Feature: RFC 0023 Establishing Connections with DID Exchange
...
+         | unknown processing error                |
+
+   @T012-RFC0023 @normal @DerivedFunctionalTest @wip
+   Scenario: Attempt to Establish a connection with DID Exchange between two agents with an explicit invitation with connection reuse
+      Given we have "2" agents
+         | name | role      |
+         | Acme | requester |
+         | Bob  | responder |
+      And "Acme" and "Bob" have an existing connection with a public DID
+      When "Bob" sends an explicit invitation wih a public DID
```
Comment (swcurran) [inline aries-test-harness/features/0023-did-exchange.feature:163]: Typo "with"?

**Comment by swcurran** [comment]
How do I test this?  I checked out the PR, (re)built the test harness and ACA-Py Main, and then ran: `./manage run -d acapy-main -t @T012-RFC0023`. The test failed with it appearing the steps have not been implemented:

```
You can implement step definitions for undefined steps with these snippets:

@given(u'"Acme" and "Bob" have an existing connection with a public DID')
def step_impl(context):
    raise NotImplementedError(u'STEP: Given "Acme" and "Bob" have an existing connection with a public DID')

@when(u'"Bob" sends an explicit invitation wih a public DID')
def step_impl(context):
    raise NotImplementedError(u'STEP: When "Bob" sends an explicit invitation wih a public DID')

@when(u'"Acme" receives the invitation while wanting to reuse an existing connection')
def step_impl(context):
    raise NotImplementedError(u'STEP: When "Acme" receives the invitation while wanting to reuse an existing connection')

```

Looks like a type in the middle one.  Not sure about the other two.

Are there other tests that are impacted by this change?

**Comment by nodlesh** [comment]
Yes, I had originally thought I would add a test just for connection Reuse. However I quickly changed to just having it embedded in the Issue Credential and Proof Tests that support DID Exchange. I'm still contemplating adding that test, or I may just remove it, not sure yet. 
This change impacts all Issue Credential V2 tests, All Proof v2 tests, and one of each of the Issue Credential V1 and Proof v1 suites. In any of those tests, if there is an existing connection, it will get reused. 

What are your thoughts on a specific Connection Reuse Test Scenario?

**Comment by swcurran** [comment]
I don't really have an opinion. I was trying it because I didn't know how else to test the PR.  Suggest that be part of the PR -- a suggestion on how to test it.

I'm curious as to why the test didn't run. Are the existing tests using the "reuse" functions, but not those specific step implementations?

**Comment by nodlesh** [comment]
For sure, I should have stated what to do to test the PR. 

Yes, there are some changes to existing step definitions that allow for connection reuse. They all stem off of the `And Acme and Bob have an existing connection` step def. If the agent can do a reuse then it happens, if not then there is support there to just continue on as usual, so no impact to other agents that run these tests.

Those steps in that new test are not implemented, and the test is tagged @wip.

**Review by swcurran — APPROVED** [review]

**Pull request merged**